### PR TITLE
(fix): correct typing of `AnnData.X`

### PR DESCRIPTION
**Pull request by ilan-gold**
- [ ] Re: https://github.com/scverse/anndata/pull/1598#issuecomment-2314890727
- [ ] Tests added
- [ ] Release note added (or unnecessary)

### CODE REVIEW

File: src/anndata/_core/anndata.py [hunk 1]
```diff
@@ -541,7 +542,17 @@ def shape(self) -> tuple[int, int]:
...
 
     @property
-    def X(self) -> np.ndarray | sparse.spmatrix | SpArray | ArrayView | None:
+    def X(
+        self,
+    ) -> (
+        CSRDataset
+        | CSCDataset
+        | np.ndarray
+        | sparse.spmatrix
```
Comment (flying-sheep) [inline src/anndata/_core/anndata.py:551]:
```suggestion
        | sparse.csr_matrix
        | sparse.csc_matrix
```

**Review by flying-sheep — COMMENTED** [review]
Shouldn’t we unify this with `ArrayDataStructureType`?

I don’t quite know why that one is an enum, maybe we can replace it with

```py
ArrayDataStructureType = Union[All|Of|This|Here]
```

and then do

```
from typing import get_args

isinstance(get_args(ArrayDataStructureType))
```

**Comment by ilan-gold** [comment]
> I don’t quite know why that one is an enum, maybe we can replace it with

I also didn't know but don't really question these sorts of things.  I imagine it served a purpose at some point, but like I said in the related PR, that seems to have changed recently.

**Comment by flying-sheep** [comment]
I guess then we should unify the two. Either as part of this PR or later, as you wish

File: src/anndata/_core/anndata.py (hunk 1 above)
Comment (ilan-gold) [reply src/anndata/_core/anndata.py:551]: So somewhere, we are sticking a `dia_matrix` into an `AnnData` object.  For now I'm going to leave this, and file an issue.

Comment (ilan-gold) [reply src/anndata/_core/anndata.py:551]: Nope! too quick! this caught a bug in a test! good!

**Comment by flying-sheep** [comment]
We could maybe use [`__instancecheck__`](https://docs.python.org/3/reference/datamodel.html#class.__instancecheck__) and [`is_array_api_obj`](https://data-apis.org/array-api-compat/helper-functions.html#array_api_compat.is_array_api_obj):

```py
import array_api_compat

class ArrayApiMeta(type):
    def __instancecheck__(cls, instance: object):
        return array_api_compat.is_array_api_obj(instance)

class ArrayApi(meta=ArrayApiMeta):
    pass
```

**Comment by ilan-gold** [comment]
I'm going to push back on the array-api here because I think it's out of scope and testing for it will be non-trivial.  I think we should make a separate branch/PR for that specifically, and stick it in `0.12`.  We can test for jax, which now has a working CPU implementation (like numpy).  This would be a very cool feature.

**Comment by ilan-gold** [comment]
There's even an issue for this: https://github.com/scverse/anndata/issues/1195

File: src/anndata/_core/storage.py [hunk 2]
```diff
@@ -28,45 +27,23 @@
...
+    np.ndarray,
+    ma.MaskedArray,
+    sparse.csr_matrix,
+    sparse.csc_matrix,
+    SpArray,
+    AwkArray,
+    H5Array,
+    ZarrArray,
+    ZappyArray,
+    BaseCompressedSparseDataset,
```
Comment (flying-sheep) [inline src/anndata/_core/storage.py:42]: Similar to above with `csr_matrix` and `csc_matrix`, we should probably do this, what do you think?

```suggestion
    CSRDataset,
    CSCDataset,
```

Comment (flying-sheep) [reply src/anndata/_core/storage.py]: self-answer: yes, we should, this way we get links.

File: src/anndata/_core/storage.py [hunk 3]
```diff
@@ -100,7 +87,7 @@ def coerce_array(
         except (ValueError, TypeError) as _e:
             e = _e
     # if value isn’t the right type or convertible, raise an error
-    msg = f"{name} needs to be of one of {join_english(ArrayDataStructureType.qualnames())}, not {type(value)}."
+    msg = f"{name} needs to be of one of {join_english([str(x) for x in array_data_structure_types])}, not {type(value)}."
```
Comment (flying-sheep) [inline src/anndata/_core/storage.py:90]:
```suggestion
    msg = f"{name} needs to be of one of {join_english(map(str, array_data_structure_types))}, not {type(value)}."
```

**Review by flying-sheep — APPROVED** [review]

**Pull request merged**